File: askmate/ingestion/notion.py
from __future__ import annotations
from typing import List
import os

from notion_client import Client  # type: ignore

from askmate.store import AskMateVectorStore
from askmate.ingestion.pdf import chunk_text


def ingest_notion_pages(database_id: str, vector_store: AskMateVectorStore) -> int:
    api_key = os.getenv("NOTION_API_KEY")
    if not api_key:
        return 0
    notion = Client(auth=api_key)
    results: List[dict] = []
    if database_id:
        query = notion.databases.query(database_id=database_id)
        results.extend(query.get("results", []))
    else:
        # Fetch recent pages as fallback
        results = notion.search(query="", page_size=10).get("results", [])

    total = 0
    for page in results:
        page_id = page.get("id")
        if not page_id:
            continue
        title = page.get("properties", {}).get("Name", {}).get("title", [])
        title_text = " ".join([t.get("plain_text", "") for t in title]) if isinstance(title, list) else str(title)
        # Pull blocks
        blocks = notion.blocks.children.list(page_id)
        texts: List[str] = []
        for b in blocks.get("results", []):
            rt = b.get(b.get("type", ""), {}).get("rich_text", [])
            txt = " ".join([t.get("plain_text", "") for t in rt]) if isinstance(rt, list) else ""
            if txt:
                texts.append(txt)
        combined = (title_text + "\n" + "\n".join(texts)).strip()
        chunks = chunk_text(combined)
        metas = [{"source": f"Notion:{title_text}", "type": "notion"} for _ in chunks]
        total += vector_store.add_texts(chunks, metas)
    return total

File: node-backend/src/ingestion/notion.ts
import { Client } from '@notionhq/client';
import { VectorStore } from '../vectorStore';

function chunkText(text: string, chunkSize = 800, overlap = 120): string[] {
  const chunks: string[] = [];
  let start = 0;
  while (start < text.length) { const end = Math.min(text.length, start + chunkSize); chunks.push(text.slice(start, end)); start = Math.max(0, end - overlap); }
  return chunks.map(c => c.trim()).filter(Boolean);
}

export async function ingestNotion(databaseId: string | undefined, store: VectorStore, apiKey: string): Promise<number> {
  const notion = new Client({ auth: apiKey });
  let pages: any[] = [];
  if (databaseId) {
    const q = await notion.databases.query({ database_id: databaseId });
    pages = q.results as any[];
  } else {
    const s = await notion.search({ page_size: 10 });
    pages = s.results as any[];
  }
  let total = 0;
  for (const page of pages) {
    const pageId = page.id;
    const props = (page as any).properties || {};
    const name = (props.Name?.title || []).map((t: any) => t.plain_text).join(' ');
    const blocks = await notion.blocks.children.list({ block_id: pageId });
    const texts: string[] = [];
    for (const b of blocks.results as any[]) {
      const t = (b[b.type]?.rich_text || []).map((rt: any) => rt.plain_text).join(' ');
      if (t) texts.push(t);
    }
    const content = (name + '\n' + texts.join('\n')).trim();
    const chunks = chunkText(content);
    const metas = chunks.map(() => ({ source: `Notion:${name}`, type: 'notion' }));
    total += await store.addTexts(chunks, metas);
  }
  return total;
}